src/megaapi_impl.cpp

```diff
@@ -18844,11 +18846,17 @@ unsigned MegaApiImpl::sendPendingTransfers()
             fireOnTransferFinish(transfer, make_unique<MegaErrorPrivate>(e), committer);
         }
 
-        if (++count > 100 || std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - t0).count() > 100)
+        auto elapsedTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - t0).count();
+        if (elapsedTime > maxPendingTransfersTimeout
+                || (!mProcessingFolderTransfer && (++count > 100 || elapsedTime > minPendingTransfersTimeout)))
         {
+            // if elapsed time exceeds 3000 ms, and we are processing a folder transfer
+            // or we're not processing a folder transfer, and processed transfers exceed 100
```

Comment:
Three seconds is a long time, especially if the app is trying to draw a window, and needs to wait for the mutex before it can draw something.     Since we are reworking folder uploads/downloads, I think this is the right time - we should perform the directory scanning on a thread (the thread can just exit once it's finished).   As it runs, it scans the folder and all subfolders, building up a data structure of the folders to create, and files in those folders to upload.   Whatever it scans is the snapshot that will actually be uploaded, so no race conditions or callback issues, or problems with that operation taking the mutex for too long - and we can keep the old 100ms condition here.